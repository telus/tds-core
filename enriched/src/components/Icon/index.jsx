import React, { PropTypes } from 'react';

function Icon(props) {
  const { glyph, variant, fixedWidth, className, children, ...extraProps } = props;

  const classes = [
    'icon',
    `icon-core-${glyph}`,
    fixedWidth ? 'icon--fw' : null,
    variant ? `icon--${variant}` : null,
    className
  ];

  return (
    <i className={classes.filter(c => c).join(' ')} {...extraProps}>
      {props.children}
    </i>
  );
}

Icon.propTypes = {
  glyph: PropTypes.string.isRequired,
  variant: PropTypes.string,
  fixedWidth: PropTypes.bool,
  className: PropTypes.string,
  children: React.PropTypes.node
};

export default Icon;
