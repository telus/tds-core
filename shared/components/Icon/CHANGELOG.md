# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

# [1.1.0](https://github.com/telus/tds-core/compare/@tds/shared-icon@1.0.1...@tds/shared-icon@1.1.0) (2019-03-15)


### Features

* **shared-icon:** add svg icon classes ([894dddf](https://github.com/telus/tds-core/commit/894dddf))





## [1.0.1](https://github.com/telus/tds-core/compare/@tds/shared-icon@1.0.0...@tds/shared-icon@1.0.1) (2019-03-05)

**Note:** Version bump only for package @tds/shared-icon





# 1.0.0 (2019-02-06)


### Features

* **colours:** create a colours package ([8f7e5d1](https://github.com/telus/tds-core/commit/8f7e5d1))
