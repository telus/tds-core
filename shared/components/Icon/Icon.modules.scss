@import '~@tds/core-colours/colours';

.primary {
  color: $color-icon-primary;
}

.secondary {
  color: $color-icon-secondary;
}

.inverted {
  color: $color-white;
}

.error {
  color: $color-cardinal;
}

.size16 {
  font-size: 1rem;
}

.size24 {
  font-size: 1.5rem;
}

.size48 {
  font-size: 3rem;
}

.svgBase {
  display: inline-flex;
}
.svgSize16 > svg {
  width: 16px;
  height: 16px;
}
.svgSize24 > svg {
  width: 24px;
  height: 24px;
}
.svgSize48 > svg {
  width: 48px;
  height: 48px;
}

.svgVariantDefault > svg {
  fill: $color-telus-purple;
}
.svgVariantAlternative > svg {
  fill: $color-shark;
}
.svgVariantInverted > svg {
  fill: $color-white;
}

.icon {
  font-family: 'TELUS Core Icons'; /* stylelint-disable-line font-family-no-missing-generic-family-keyword */

  display: inline-block;
  font-weight: normal;
  font-style: normal;
  speak: none;
  text-decoration: inherit;
  text-transform: none;
  text-rendering: auto;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;

  line-height: 1;
  vertical-align: middle;
}

.iconCaretDown {
  composes: icon;

  &::before {
    content: '\f105';
  }
}

.iconCaretUp {
  composes: icon;

  &::before {
    content: '\f106';
  }
}

.iconCheckmark {
  composes: icon;

  &::before {
    content: '\f101';
  }
}

.iconChevron {
  composes: icon;

  &::before {
    content: '\f107';
  }
}

.iconLeftChevron {
  composes: icon;

  &::before {
    content: '\f107';
    display: inline-block;
    transform: rotate(-180deg) translateY(1.5px);
  }
}

.iconExclamationPointCircle {
  composes: icon;

  &::before {
    content: '\f103';
  }
}

.iconExpander {
  composes: icon;

  &::before {
    content: '\f113';
  }
}

.iconHamburger {
  composes: icon;

  &::before {
    content: '\f112';
  }
}

.iconLocation {
  composes: icon;

  &::before {
    content: '\f110';
  }
}

.iconMinus {
  composes: icon;

  &::before {
    content: '\f109';
  }
}

.iconPlus {
  composes: icon;

  &::before {
    content: '\f108';
  }
}

.iconQuestionMarkCircle {
  composes: icon;

  &::before {
    content: '\f102';
  }
}

.iconSpyglass {
  composes: icon;

  &::before {
    content: '\f111';
  }
}

.iconTimes {
  composes: icon;

  &::before {
    content: '\f104';
  }
}
