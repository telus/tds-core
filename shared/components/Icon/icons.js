const icons = {
  default: {
    fontFamily:
      "'TELUS Core Icons'" /* stylelint-disable-line font-family-no-missing-generic-family-keyword */,
    display: 'inline-block',
    fontWeight: 'normal',
    fontStyle: 'normal',
    speak: 'none',
    textDecoration: 'inherit',
    textTransform: 'none',
    textRendering: 'auto',
    '-webkit-font-smoothing': 'antialiased',
    '-moz-osx-font-smoothing': 'grayscale',
    lineHeight: 1,
    verticalAlign: 'middle',
  },
  caretDown: {
    '&::before': {
      content: "'\f105'",
    },
  },
  caretUp: {
    '&::before': {
      content: "'\f106'",
    },
  },
  checkmark: {
    '&::before': {
      content: "'\f101'",
    },
  },
  chevron: {
    '&::before': {
      content: "'\f107'",
    },
  },
  leftChevron: {
    '&::before': {
      content: "'\f107'",
      display: 'inline-block',
      transform: 'rotate(-180deg) translateY(1.5px)',
    },
  },
  exclamationPointCircle: {
    '&::before': {
      content: "'\f103'",
    },
  },
  expander: {
    '&::before': {
      content: "'\f113'",
    },
  },
  hamburger: {
    '&::before': {
      content: "'\f112'",
    },
  },
  location: {
    '&::before': {
      content: "'\f110'",
    },
  },
  minus: {
    '&::before': {
      content: "'\f109'",
    },
  },
  plus: {
    '&::before': {
      content: "'\f108'",
    },
  },
  questionMarkCircle: {
    '&::before': {
      content: "'\f102'",
    },
  },
  spyglass: {
    '&::before': {
      content: "'\f111'",
    },
  },
  times: {
    '&::before': {
      content: "'\f104'",
    },
  },
}
export default icons
