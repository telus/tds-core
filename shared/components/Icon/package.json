{
  "name": "@tds/shared-icon",
  "version": "1.1.3",
  "private": true,
  "main": "",
  "module": "",
  "peerDependencies": {
    "react": "^16.8.2",
    "react-dom": "^16.8.2",
    "styled-components": "^4.1.3"
  },
  "dependencies": {
    "@tds/core-colours": "^1.1.2"
  }
}
