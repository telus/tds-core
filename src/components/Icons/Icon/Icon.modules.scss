@import '../../../scss/settings/colours';

// Unicode points of each icon's location within the font files.
$core-icon-codepoints: (
  alert: '\f103',
  caret-down: '\f105',
  caret-up: '\f106',
  checkmark: '\f101',
  chevron: '\f107',
  close: '\f104',
  dash: '\f109',
  exclamation-point-circle: '\f103',
  expander: '\f113',
  hamburger: '\f112',
  help: '\f102',
  incomplete: '\f104',
  location: '\f110',
  minus: '\f109',
  plus: '\f108',
  question-mark-circle: '\f102',
  spyglass: '\f111',
  times: '\f104'
);

// Get the unicode point for a given icon name
@function core-icon-codepoint($name) {
  @return map-get($core-icon-codepoints, $name);
}

@mixin core-icon($name: null) {
  @if $name {
    &::before {
      content: core-icon-codepoint($name);
    }
  }

  font-family: 'TELUS Core Icons';
  display: inline-block;
  font-weight: normal;
  font-style: normal;
  speak: none;
  text-decoration: inherit;
  text-transform: none;
  text-rendering: auto;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
}

.primary {
  color: $color-icon-primary;
}

.secondary {
  color: $color-icon-secondary;
}

.inverted {
  color: $color-white;
}

.error {
  color: $color-cardinal;
}

.size16 {
  font-size: 1rem;
}

.size24 {
  font-size: 1.5rem;
}

.size48 {
  font-size: 3rem;
}

.icon {
  @include core-icon;

  line-height: 1;
  vertical-align: middle;
}

.iconCaretDown {
  composes: icon;

  &::before {
    content: '\f105';
  }
}

.iconCaretUp {
  composes: icon;

  &::before {
    content: '\f106';
  }
}

.iconCheckmark {
  composes: icon;

  &::before {
    content: '\f101';
  }
}

.iconChevron {
  composes: icon;

  &::before {
    content: '\f107';
  }
}

.iconLeftChevron {
  composes: icon;

  &::before {
    content: '\f107';
    display: inline-block;
    transform: rotate(-180deg) translateY(1.5px);
  }
}

.iconExclamationPointCircle {
  composes: icon;

  &::before {
    content: '\f103';
  }
}

.iconExpander {
  composes: icon;

  &::before {
    content: '\f113';
  }
}

.iconHamburger {
  composes: icon;

  &::before {
    content: '\f112';
  }
}

.iconLocation {
  composes: icon;

  &::before {
    content: '\f110';
  }
}

.iconMinus {
  composes: icon;

  &::before {
    content: '\f109';
  }
}

.iconPlus {
  composes: icon;

  &::before {
    content: '\f108';
  }
}

.iconQuestionMarkCircle {
  composes: icon;

  &::before {
    content: '\f102';
  }
}

.iconSpyglass {
  composes: icon;

  &::before {
    content: '\f111';
  }
}

.iconTimes {
  composes: icon;

  &::before {
    content: '\f104';
  }
}
