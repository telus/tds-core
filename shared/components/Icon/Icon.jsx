import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'

import { colorIconPrimary, colorIconSecondary, colorCardinal, colorWhite } from '@tds/core-colours'

import icons from './icons'

import safeRest from '../../utils/safeRest'

const getColour = variant => {
  switch (variant) {
    case 'primary':
      return colorIconPrimary
    case 'secondary':
      return colorIconSecondary
    case 'inverted':
      return colorWhite
    case 'error':
      return colorCardinal
    default:
      return undefined
  }
}
const iconSymbol = ({ symbol }) => ({ ...icons.default, ...icons[symbol] })
const iconVariant = ({ variant }) => ({ color: getColour(variant) })
const iconSize = ({ iSize }) => ({ fontSize: `${iSize / 16}rem` })

export const StyledIcon = styled.i(iconSymbol, iconVariant, iconSize)

const Icon = ({ symbol, variant, size, ...rest }) => (
  <StyledIcon {...safeRest(rest)} symbol={symbol} variant={variant} iSize={size} />
)

Icon.propTypes = {
  symbol: PropTypes.oneOf([
    'caretDown',
    'caretUp',
    'checkmark',
    'chevron',
    'leftChevron',
    'exclamationPointCircle',
    'expander',
    'hamburger',
    'location',
    'minus',
    'plus',
    'questionMarkCircle',
    'spyglass',
    'times',
  ]).isRequired,
  variant: PropTypes.oneOf(['primary', 'secondary', 'inverted', 'error']),
  size: PropTypes.oneOf([16, 24, 48]),
}

Icon.defaultProps = {
  variant: undefined,
  size: 24,
}

export default Icon
