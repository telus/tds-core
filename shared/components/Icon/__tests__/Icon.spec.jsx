import React from 'react'
import { shallow } from 'enzyme'

import { colorIconSecondary } from '@tds/core-colours'

import Icon, { StyledIcon } from '../Icon'

describe('Icon', () => {
  const defaultProps = {
    symbol: 'spyglass',
  }
  const doShallow = (props = {}) => shallow(<Icon {...defaultProps} {...props} />)
  const doShallowStyled = (props = {}) => shallow(<StyledIcon {...defaultProps} {...props} />)

  it('renders', () => {
    const icon = doShallow({ symbol: 'spyglass' })
    expect(icon).toMatchSnapshot()
  })

  it('supports variants', () => {
    const icon = doShallowStyled({ variant: 'secondary' })

    expect(icon).toHaveStyleRule('color', colorIconSecondary)
  })

  it('can be sized', () => {
    const icon = doShallowStyled({ iSize: 16 })

    expect(icon).toHaveStyleRule('font-size', '1rem')
  })

  it('passes additional attributes to the i element', () => {
    const icon = doShallow({ id: 'the-icon', role: 'button' })

    expect(icon).toHaveProp('id', 'the-icon')
    expect(icon).toHaveProp('role', 'button')
  })

  it('does not allow custom CSS', () => {
    const icon = doShallow({ className: 'my-custom-class', style: { color: 'hotpink' } })

    expect(icon).not.toHaveProp('className', 'my-custom-class')
    expect(icon).not.toHaveProp('style')
  })
})
