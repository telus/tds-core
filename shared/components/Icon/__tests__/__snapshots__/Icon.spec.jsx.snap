// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Icon renders 1`] = `
<Icon__StyledIcon
  iSize={24}
  symbol="spyglass"
/>
`;
